src/libraries/System.Private.CoreLib/src/System/IO/Directory.Windows.cs

```diff
@@ -1,11 +1,68 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Runtime.InteropServices;
+using System.Text;
+
 namespace System.IO
 {
     public static partial class Directory
     {
+        private static readonly char[] s_directorySeparators = new char[] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar };
+
         private static DirectoryInfo CreateDirectoryCore(string path, UnixFileMode unixCreateMode)
             => throw new PlatformNotSupportedException(SR.PlatformNotSupported_UnixFileMode);
+
+        private static unsafe string CreateTempSubdirectoryCore(string? prefix)
+        {
+            ValueStringBuilder builder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
+            Path.GetTempPath(ref builder);
+
+            // ensure the base TEMP directory exists
+            CreateDirectory(PathHelper.Normalize(ref builder));
+
+            builder.Append(prefix);
+
+            const int RandomFileNameLength = 12; // 12 == 8 + 1 (for period) + 3
+            int initialTempPathLength = builder.Length;
+            builder.EnsureCapacity(initialTempPathLength + RandomFileNameLength);
+
+            // For generating random file names
+            // 8 random bytes provides 12 chars in our encoding for the 8.3 name.
+            const int RandomKeyLength = 8;
+            byte* pKey = stackalloc byte[RandomKeyLength];
+
+            // to avoid an infinite loop, only try as many as GetTempFileNameW will create
+            const int MaxAttempts = ushort.MaxValue;
+            int attempts = 0;
+            while (attempts < MaxAttempts)
+            {
+                Interop.GetRandomBytes(pKey, RandomKeyLength);
+                Path.Populate83FileNameFromRandomBytes(pKey, RandomKeyLength, builder.RawChars.Slice(builder.Length, RandomFileNameLength));
```

Comment:
Why is it important to use the 8.3 support?

Reply:
It isn't important. But it is what GetRandomFileName uses

https://github.com/dotnet/runtime/blob/916f1ad5b391764e964b271f80b915869e1afed0/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs#L281-L289

I'm just simulating a call to `GetRandomFileName` here, without allocating an intermediate string.

Reply:
I can add a comment here, if that helps. Or did you have a different algorithm in mind for generating a folder name?

Reply:
IMO having a period in the directory name seems a bit strange

Reply:
> IMO having a period in the directory name seems a bit strange

Does it really matter though? Code shouldn't depend on these names - they even have a different pattern between Windows and Unix. Besides, there are plenty of cases where directories have periods in their names - versions, namespace-like, etc.

e.g. `C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.7`.